ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,501 @@
+
+      // check for case D described in the javadoc of AccountService
+      if (remoteAccount == null && currentAccounts.containsName(account.getName())) {
+        Account conflictingAccount = currentAccounts.getAccountByName(account.getName());
+        logger.error(
+            "Account to update with accountId={} accountName={} conflicts with an existing record with accountId={} accountName={}",
+            account.getId(), account.getName(), conflictingAccount.getId(), conflictingAccount.getName());
+        res = true;
+        break;
+      }
+
+      // check for case E described in the javadoc of AccountService
+      if (remoteAccount != null && !account.getName().equals(remoteAccount.getName()) && currentAccounts.containsName(
+          account.getName())) {
+        Account conflictingAccount = currentAccounts.getAccountByName(account.getName());
+        logger.error(
+            "Account to update with accountId={} accountName={} conflicts with an existing record with accountId={} accountName={}",
+            account.getId(), account.getName(), conflictingAccount.getId(), conflictingAccount.getName());
+        res = true;
+        break;
+      }
+    }
+    return res;
+  }
+
+  /**
+   * Checks if the {@code HelixAccountService} is open.
+   */
+  private void checkOpen() {
+    if (!isOpen) {
+      throw new IllegalStateException("AccountService is closed.");
+    }
+  }
+
+  /**
+   * <p>
+   *   A helper class that represents a collection of {@link Account}s, where the ids and names of the
+   *   {@link Account}s are one-to-one mapped. An {@code AccountInfoMap} guarantees no duplicated account
+   *   id or name, nor conflict among the {@link Account}s within it.
+   * </p>
+   * <p>
+   *   Based on the properties, a {@code AccountInfoMap} internally builds index for {@link Account}s using both
+   *   {@link Account}'s id and name as key.
+   * </p>
+   */
+  private class AccountInfoMap {
+    private final Map<String, Account> nameToAccountMap;
+    private final Map<Short, Account> idToAccountMap;
+
+    /**
+     * Constructor for an empty {@code AccountInfoMap}.
+     */
+    private AccountInfoMap() {
+      nameToAccountMap = new HashMap<>();
+      idToAccountMap = new HashMap<>();
+    }
+
+    /**
+     * <p>
+     *   Constructs an {@code AccountInfoMap} from a group of {@link Account}s. The {@link Account}s exists
+     *   in the form of a string-to-string map, where the key is the string form of an {@link Account}'s id,
+     *   and the value is the string form of the {@link Account}'s JSON string.
+     * </p>
+     * <p>
+     *   The source {@link Account}s in the {@code accountMap} may have duplicate account Ids or names. The
+     *   constructor makes the best effort to resolve the conflict. Based on the order when iterating on the
+     *   {@link Account}s, the constructor will: 1) override the previously-added {@link Account} if a latter
```

Comment:
Have updated the javadoc. Explains no `AccountInfoMap` could be constructed if there is any conflict.